cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,376 @@
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+        n_samples: number of runs for sampling
+
+    Returns:
+        expectation value of operator argument
+
+    """
+
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
+
+    # remove measurements from circuit such that appropriate rotations
+    # can be applied, measurements are added later
+    circuit = no_meas(circuit)
+
+    # check if operator is quadratic in Pauli Z:
+    quadratic_z = True
+    zz = [Pauli.Z, Pauli.Z]
+    for key in operator.keys():
+        if (len(key) != 2):
+            quadratic_z = False
+            break
+        if list(key.values()) != zz:
+            quadratic_z = False
+            break
+
+    # if operator is only of the form ZZ then call helper function
+    if quadratic_z:
+
+        expectation = _expectation_from_sampling_assuming_quadratic(circuit,
+                                                                    operator,
+                                                                    n_samples)
+        return expectation
+
+
+    # For general operators, the expectation value by sampling
+    # over n_samples circuit runs is computed here
+    if not quadratic_z:
+
+        expectation = 0
+        identity_coeficient = 0.
+
+        for term, coef in operator.items():
+
+            # remove old measurements
+            circuit = no_meas(circuit)
+
+            # check if identity:
+            if term == ():
+                identity_coeficient = coef
+                continue
+
+            # Add appropriate rotation gates to circuit
+            # and add measurement gates
+            # dictionary of qubits to be measured and its indices
+            q_dict = dict()
+
+            for qubit, pauli in term.items():
+
+                if pauli == Pauli.X:
+                    circuit.append(RotYGate(half_turns=-1 / 2).on(
+                        qubit))
+
+                elif pauli == Pauli.Y:
+                    circuit.append(RotXGate(half_turns=+1 / 2).on(
+                        qubit))
+
+                # After rotation we can re-add measurement gates
+                q_dict[qubit] = qubits.index(qubit)
+
+            # Add Measurement gates:
+            circuit = measurement_gates(circuit, q_dict)
+
+            # run circuit:
+            results = sim.run(circuit, repetitions=n_samples)
+
+            operator_meas = 1
+
+            for qubit_key in results.measurements.keys():
+                operator_meas *= results.measurements[qubit_key][:, 0] \
```

Comment:
Nit: style guide is to do line continuations using parens instead of `\`. Occurs in several places.